lib/logic/box_communicator.dart

```diff
@@ -81,6 +87,110 @@ class BoxCommunicator {
+    List idList = List();
+    Map<String, String> headers = {
+      "Content-type": "application/json",
+    };
+
+    //register to get the current boxName and get a id(timestamp) from the Sensorbox
+    final boxResponse = await retry(
+        () => http.get(boxIP + "/api/register").timeout(Duration(seconds: 3)),
+        retryIf: (e) => e is SocketException || e is TimeoutException);
+    if (boxResponse.statusCode == 200) {
+      var res = jsonDecode(boxResponse.body);
+      boxName = res["piID"];
+      deviceTimestamp = res["timestamp"];
+      //storage box stores the number of received bytes at 'totalSizeInBytes'
+      //opens the box for the current connected Sensorbox
+      if (box == null) {
+        box = await Hive.openLazyBox(boxName);
+      }
+      print("Opened box with boxname " + boxName.toString());
+      //If List of IDs is empty execute the download without a data limit and without priority restrictions
+      if (box.keys.toList().length > 0) {
+        idList = box.keys.toList();
+      }
+      //after getting list of ids on device for corresponding boxname we send the list to the box among device id (timestamp)
+      IdListAndTimeStamp idListAndTimeStamp =
+          new IdListAndTimeStamp(timestamp: deviceTimestamp, idList: idList);
+      String body = json.encode(idListAndTimeStamp);
+      final currentDataRegisterRespone = await retry(
+          () => http
+              .post(boxIP + "/api/registerCurrentData",
+                  headers: headers, body: body)
+              .timeout(Duration(seconds: 3)),
+          retryIf: (e) => e is SocketException || e is TimeoutException);
+      // if the files db AND chunk db is empty we get a notification and the download stops
+      if (currentDataRegisterRespone.statusCode == 201) {
+        print("No data on box");
+        context.read<DownloadAllState>().completed();
+        await Future.delayed(const Duration(seconds: 15));
+        context.read<DownloadAllState>().initial();
+        return;
+      }
+      if (currentDataRegisterRespone.statusCode == 200) {
+        print("Box recieved ids and timestamp successfully");
+      } else {
+        throw Exception('Failed to send IDs and timestamp with status code ' +
+            currentDataRegisterRespone.statusCode.toString());
+      }
+    } else {
+      throw Exception('failed to register');
+    }
+
+    //if no data has been received: start with 0.
+    int totalSizeInBytes = storage.get('totalSizeInBytes', defaultValue: 0);
+    //now this loop will be executed when the general download loop is paused.
+    // Rest of work is done on the box
+    while (!_executeDownload) {
+      IdListAndTimeStamp idListAndTimeStamp =
+          new IdListAndTimeStamp(timestamp: deviceTimestamp, idList: List());
+      String body = json.encode(idListAndTimeStamp);
+      final response = await retry(
+          () => http
+              .post(boxIP + "/api/getAllData", headers: headers, body: body)
```

Comment:
why is this a post request?

Reply:
http.get does not support bodies

Reply:
but you could pass parameters in url (see fetch Image by ID)

Reply:
Done